### PR TITLE
Move Container to its own package.

### PR DESCRIPTION
**Pull request by calavera**
Some people will like it, a lot of people will hate it, but we need to do this to stop importing all the daemon every time we need a container.

Sorry, but I've been fixing compilation errors for two days and I couldn't find a sane way of dividing this into smaller commits.

/cc @tiborvass

Signed-off-by: David Calavera david.calavera@gmail.com

### CODE REVIEW

**Comment by cpuguy83** [comment]
```
daemon/container_operations_unix.go:161: c.hostConfig undefined (type *container.Container has no field or method hostConfig)
daemon/exec_unix.go:12: undefined: Container
```

**Comment by calavera** [comment]
yep, dealing with some merge conflicts now :sparkles:

**Comment by cpuguy83** [comment]
Also +1. Wondering if a follow-on to this should factor out libnetwork from the container package?

**Comment by calavera** [comment]
updated, at least, it should compile now :grimacing:

**Comment by duglin** [comment]
Still making my way thru this but I think the general direction looks good, mainly because I like the idea of containers be more self-defined. But aren't we getting into kind of a weird state because while you can import containers w/o the daemon, in order to do most of the actions on a container you'll need the daemon?  e.g. ContainerStop() is on Daemon not Container, so people might still need to import daemon anyway to actually manipulate it.  Can we look into containers being even more independent/self-defined by having them also own the funcs that operate on containers too?

**Comment by MHBauer** [comment]
I'm taking a deep look at this as well. Definitely think it is a good concept!

File: container/container_unix.go [hunk 1]
```diff
@@ -0,0 +1,683 @@
...
+			Source:      container.MqueuePath,
+			Destination: "/dev/mqueue",
+			Writable:    true,
+			Private:     true,
+		})
+	}
+	return mounts
+}
+
+// UnmountVolumes explicitely unmounts volumes from the container.
```
Comment (jhowardmsft) [inline container/container_unix.go]: Typo: explicitly

File: container/container_windows.go [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+// This is a NOOP on windows.
+func (container *Container) UnmountIpcMounts(unmount func(pth string) error) {
+}
+
+// IpcMounts returns the list of Ipc related mounts.
+func (container *Container) IpcMounts() []execdriver.Mount {
+	return nil
+}
+
+// UnmountVolumes explicitely unmounts volumes from the container.
```
Comment (jhowardmsft) [inline container/container_windows.go]: Typo.  Can we also copy across the comment from daemon\container_windows.go for fixing post TP4 (that comment applies to both mountVolumes and unmountVolumes). Thanks.

File: daemon/exec/exec.go [hunk 3]
```diff
@@ -0,0 +1,110 @@
+package exec
```
Comment (MHBauer) [inline daemon/exec/exec.go]: This packaged was created and exec was moved to avoid a dependency on daemon, by container.

**Comment by MHBauer** [comment]
Hard to review in it's entirety. Do you have a quick estimate on how much is moved/renamed for export vs how much new code? It looks like the vast majority is moving stuff around, but I want to make sure to look at anything new for sure.

File: container/archive.go [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+	}
+
+	// resolvedDirPath will have been cleaned (no trailing path separators) so
+	// we can manually join it with the base path element.
+	resolvedPath = resolvedDirPath + string(filepath.Separator) + basePath
+
+	return resolvedPath, absPath, nil
+}
+
+// StatPath is the unexported version of StatPath. Locks and mounts should
```
Comment (calavera) [inline container/archive.go]: There is another function that calls this one holding a lock. I'd need to double check to make sure this one is not called from anywhere else, but I had to export it to make everything compile successfully, so my guess is that it's called somewhere else. I really didn't look into callers.

File: daemon/exec/exec.go (hunk 3 above)
Comment (calavera) [inline daemon/exec/exec.go]: yes, I created it to remove a cyclic dependency.

**Comment by calavera** [comment]
@MHBauer good question. I did not add any new code that was not in master already. There are three new packages that I created to remove cycled dependencies: `container`, `container/streams` and `daemon/exec`. I tried to minimize the changes to only move functions and fields from unexported to exported.

**Comment by calavera** [comment]
Closing temporarily. There are a few things that we extract cleanly from here. There will be less things to review when that happens.

Please review #18051 meanwhile.

**Comment by MHBauer** [comment]
@calavera Anything I can help to extract?

**Comment by duglin** [comment]
@calavera I'd like to get your take on: https://github.com/docker/docker/pull/17969#issuecomment-156502395

**Comment by tiborvass** [comment]
I want this now

**Pull request closed without merging**